### PR TITLE
[Internal Cleanup] pre-commit ruff (excluding docs/tools, libcudacxx/test)

### PR DESCRIPTION
**Pull request by rwgk**
Consolidation of results obtained in prior incremental PRs into 3 commits:

1. Add ruff configs: 4114ee25dd2da7ddd0d15bad17d89f29530c6c68 (**Please review**)
2. ruff auto-fixes (NO manual changes): 9413cafa3a377b36d835ae6338feb2dd181c1677 (**Do not review**)
3. All manual fixes: c716ce3c7bbd4df1adf967c33f33072a14b7d8be (**Please review**)

The cleanup uncovered these oversights:

* benchmarks/scripts/analyze.py:

```diff
     if alpha < 1.0:                                                            
-        case_df = remove_matching_distributions(alpha, case_df)                
+        for subbench in case_dfs:                                              
+            case_dfs[subbench] = remove_matching_distributions(                
+                alpha, case_dfs[subbench]                                      
+            )                                                                  
```

* cub/benchmarks/docker/recipe.py:

```diff
-Stage0 += hpccm.primitives.baseimage(image='nvidia/cuda:12.2.0-devel-ubuntu22.04')
+Stage0 = hpccm.primitives.baseimage(image="nvidia/cuda:12.2.0-devel-ubuntu22.04")
```

________
For easy future reference, the list of prior incremental PRs (all obsolete): #3104, #3101, #3099, #3086

### CODE REVIEW

**Comment by bernhardmgruber** [comment]
I think it's great to have some autoformatting (and linting) for python code! What's most important to me is that pre-commit and the CI can handle and check that for me. I assume this is the case?

Regarding code style, I have no experience what's available and what's best practice. Looking at the diff, I prefer code to be dense if possible (less line breaks and less empty lines), but otherwise I have no opinions.

Thank you for engaging in this!

**Comment by bernhardmgruber** [comment]
```diff
-        case_df = remove_matching_distributions(alpha, case_df)                
+        remove_matching_distributions(alpha, case_dfs)  # TODO(rwgk): CORRECT? 
```
Using the function definition:
```py
def remove_matching_distributions(alpha, df):
    closure = functools.partial(distributions_are_different, alpha)
    return df[df.apply(closure, axis=1)]
```
I don't think the diff is correct, because `case_df` should refer to the new reduced data frame returned by `remove_matching_distributions`.

**Comment by rwgk** [comment]
> I think it's great to have some autoformatting (and linting) for python code! What's most important to me is that pre-commit and the CI can handle and check that for me. I assume this is the case?

Yes. This simply hooks into the existing pre-commit. You can run locally (usually I use ``), and we have the pre-commit.ci workflow in GHA, which blocks merging if it doesn't pass.

> Regarding code style, I have no experience what's available and what's best practice. Looking at the diff, I prefer code to be dense if possible (less line breaks and less empty lines), but otherwise I have no opinions.

I'm striving to keeping the defaults as much as possible. — ruff more-or-less supersedes black. black deliberately made it hard for users to be creative with style options.

File: docs/tools/packman/bootstrap/install_package.py
Comment (jrhemstad) [inline docs/tools/packman/bootstrap/install_package.py:89]: We shouldn't be formatting anything in `packman` because this is vendored code.

File: docs/tools/repoman/repoman.py
Comment (jrhemstad) [inline docs/tools/repoman/repoman.py:3]: Same with repoman.

File: .pre-commit-config.yaml
Comment (bdice) [inline .pre-commit-config.yaml:42]: Latest is v0.8.2.
```suggestion
    rev: v0.8.2
```

File: libcudacxx/test/support/filesystem_dynamic_test_helper.py
Comment (jrhemstad) [inline libcudacxx/test/support/filesystem_dynamic_test_helper.py:6]: I'm also unsure if we should be mucking with libcu++'s testing scripts. 

@wmaxey

File: pyproject.toml
Comment (bdice) [inline pyproject.toml:2]: What is the minimum version that CCCL aims to support? That is what should be used here.

I would encourage 3.10 (minimum for RAPIDS, [minimum for NEP29](https://numpy.org/neps/nep-0029-deprecation_policy.html)).

Other possibilities:
- Python 3.9 is [still receiving security updates until October 2025](https://endoflife.date/python)
- Python 3.11 is the [minimum for SPEC 0](https://scientific-python.org/specs/spec-0000/) but I feel that is too aggressive

```suggestion
target-version = "py310"
```

**Review by bdice — COMMENTED** [review]
I reviewed the two commits requested:

[Ruff configs](https://github.com/NVIDIA/cccl/commit/2c063164e435da75283cb92ba256297bcfb81e66) are mostly good, I have two suggestions.

[Manual changes](https://github.com/NVIDIA/cccl/commit/c16710fbc7f36952508da45b6b4c3ebf88802dc9) in look good to me.

The one concerning pattern here is that there are many instances of `from X import Y as Y`. The `as Y` is redundant and should be removed.

File: docs/tools/packman/bootstrap/install_package.py
Comment (bdice) [reply docs/tools/packman/bootstrap/install_package.py:89]: If it is helpful, this is how we exclude files in RAPIDS: [pyproject.toml](https://github.com/rapidsai/cudf/blob/cd3a79bfa71be68c8e95ff8dd60a41eb641f8d5a/pyproject.toml#L118-L121)

File: libcudacxx/test/support/filesystem_dynamic_test_helper.py
Comment (bdice) [reply libcudacxx/test/support/filesystem_dynamic_test_helper.py:6]: If it is helpful in deciding, I have extremely high confidence in the correctness of the rewrites that `ruff` and other linters use. The number of invalid changes they make is vanishingly small. I looked at the changes in this file and a few others in `libcudacxx/`, they seem valid to me.

**Comment by rwgk** [comment]
@bdice wrote:

> The one concerning pattern here is that there are many instances of `from X import Y as Y`. The `as Y` is redundant and should be removed.

The ruff error is e.g.:

```
benchmarks/scripts/cccl/__init__.py:1:15: F401 `.bench` imported but unused; consider removing, adding to `__all__`, or using a redundant alias
  |
1 | from . import bench
  |               ^^^^^ F401
  |
  = help: Use an explicit re-export: `bench as bench`

Found 1 error.
```

I asked ChatGPT about it: https://docs.google.com/document/d/1VoBEoLOLuKLdi2zSuXm94_EEqkIyNzSrKLVLmX_4p8w/edit?usp=sharing

`from . import bench as bench` seems like the nicest solution to me, but we could also suppress the error, either locally or globally, or use `__all__`. WDYT?

**Comment by bdice** [comment]
I recommend these solutions for F401:
- remove the import if it is actually unused
- if it is intended to be re-exported as an API for other consumers to get transitively through this module’s public API, define it in `__all__`
- add `# noqa: F401` if it is required but the static checker is unable to see its usage and thus thinks it is unused — this is also a clearer solution than `import Y as Y`, which I think will be non-obvious to most readers

**Comment by rwgk** [comment]
> I recommend these solutions for F401, in order of preference:
> 
> * remove the import if it is actually unused
> * if it is intended to be re-exported as an API for other consumers to get transitively through this module’s public API, define it in `__all__`
> * add `# noqa: F401` locally if it is required but the static checker is unable to see its usage and thus thinks it is unused — this is also a clearer solution than `import Y as Y`, which I think will be non-obvious to most readers
> * global ignores of F401 are not good, I would not recommend this

@bdice what do you think about this? — https://github.com/NVIDIA/cccl/pull/3130/commits/a4eab4bfcfa455ed72d59fc7177a1d84034e4c5d

(That commit is in a scratch PR, because I'm force-pushing to this PR, to keep the organization into 3 commits as explained in the PR description.)

File: pyproject.toml
Comment (rwgk) [reply pyproject.toml:2]: Done: https://github.com/NVIDIA/cccl/pull/3130/commits/8ac225484c96964e864644a587a0eb53979a7d01

It didn't change anything. TBH I'm still not sure if the pyproject.toml file actually has any effect for ruff running via pre-commit. E.g. putting the exclusions here definitely didn't work.

File: .pre-commit-config.yaml
Comment (rwgk) [reply .pre-commit-config.yaml:42]: Done: https://github.com/NVIDIA/cccl/pull/3130/commits/8ac225484c96964e864644a587a0eb53979a7d01

(Currently in a scratch PR; once it's settled there I'll force push here.)

File: docs/tools/packman/bootstrap/install_package.py
Comment (rwgk) [reply docs/tools/packman/bootstrap/install_package.py:89]: Will exclude (currently testing here: https://github.com/NVIDIA/cccl/pull/3130/commits/8ac225484c96964e864644a587a0eb53979a7d01
)

File: docs/tools/repoman/repoman.py
Comment (rwgk) [reply docs/tools/repoman/repoman.py:3]: Will exclude (currently testing here: https://github.com/NVIDIA/cccl/pull/3130/commits/8ac225484c96964e864644a587a0eb53979a7d01
)

File: libcudacxx/test/support/filesystem_dynamic_test_helper.py
Comment (rwgk) [reply libcudacxx/test/support/filesystem_dynamic_test_helper.py:6]: I'll exclude this, too. We can work on this in a follow-on PR, as needed.)

File: pyproject.toml
Comment (bdice) [reply pyproject.toml]: Usually the changes between Python versions are minimal-to-none.

However, I'm surprised if `pyproject.toml` settings aren't taking effect. They definitely should, from what we see in RAPIDS. You can see the [.pre-commit-config.yaml](https://github.com/rapidsai/cudf/blob/branch-25.02/.pre-commit-config.yaml) and [pyproject.toml](https://github.com/rapidsai/cudf/blob/branch-25.02/pyproject.toml#L25-L133) for an example.

Did you put the exclusions under `[tool.ruff.lint]` or `[tool.ruff]`? They should be in the `[tool.ruff.lint]` key to take effect.

**Comment by bdice** [comment]
> @bdice what do you think about this? — https://github.com/NVIDIA/cccl/commit/a4eab4bfcfa455ed72d59fc7177a1d84034e4c5d

Yes, that's fine!

And thanks for keeping the commits organized here. Normally force-pushing is harmful to reviews, but it does make it easier when working with massive linter changesets.

**Comment by rwgk** [comment]
@jrhemstad

Based on testing under #3130 this PR should be good for merging now.

Compared to what you reviewed before:

* Please re-review the very small commit for the config additions: 4dc2f79eecd35b7a847453647fc4d66f7157a821
* docs/tools and libcudacxx/test are now excluded
* the only changes to the manual fixes are https://github.com/NVIDIA/cccl/pull/3130/commits/a4eab4bfcfa455ed72d59fc7177a1d84034e4c5d, which @bdice reviewed already (thanks!).

I expect that the only test failures here will be resolved by @pciolkosz 's PR #3129. — I'm actually testing that under #3130, it'll probably be done in ~1 hour; I'd be very surprised if the testing there doesn't finish cleanly.

**Comment by rwgk** [comment]
Testing under PR #3130, which is this PR + PR #3129, passed: https://github.com/NVIDIA/cccl/pull/3130#issuecomment-2537575861

**Review by shwina — APPROVED** [review]

**Review by bdice — APPROVED** [review]

File: libcudacxx/test/support/filesystem_dynamic_test_helper.py
Comment (wmaxey) [reply libcudacxx/test/support/filesystem_dynamic_test_helper.py]: Please do! The formatting in those scripts is abysmal. :)

Comment (rwgk) [reply libcudacxx/test/support/filesystem_dynamic_test_helper.py]: OK, thanks! I'll work on that as soon as this PR is merged. It should be easy because I have the fixes already, although there was a weird "old" MSVC error that I'd need to debug if it still pops up.

**Review by leofang — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Comment by rwgk** [comment]
Thanks everybody! :-) :-)

**Pull request merged**
